crates/nu-cli/tests/commands/sum.rs

```diff
@@ -60,3 +60,162 @@ fn outputs_zero_with_no_input() {
         assert_eq!(actual.out, "0");
     })
 }
+
+#[test]
+fn computes_sum_of_individual_row() {
+    Playground::setup("sum_test_3", |dirs, sandbox| {
+        sandbox.with_files(vec![FileWithContentToBeTrimmed(
+            "calendar.json",
+            r#"
+                    [
+                        {
+                            "sunday": null,
+                            "monday": null,
+                            "tuesday": null,
+                            "wednesday": null,
+                            "thursday": null,
+                            "friday": 1,
+                            "saturday": 2
+                        },
+                        {
+                            "sunday": 3,
+                            "monday": 4,
+                            "tuesday": 5,
+                            "wednesday": 6,
+                            "thursday": 7,
+                            "friday": 8,
+                            "saturday": 9
+                        },
+                        {
+                            "sunday": 10,
+                            "monday": 11,
+                            "tuesday": 12,
+                            "wednesday": 13,
+                            "thursday": 14,
+                            "friday": 15,
+                            "saturday": 16
+                        },
+                        {
+                            "sunday": 17,
+                            "monday": 18,
+                            "tuesday": 19,
+                            "wednesday": 20,
+                            "thursday": 21,
+                            "friday": 22,
+                            "saturday": 23
+                        },
+                        {
+                            "sunday": 24,
+                            "monday": 25,
+                            "tuesday": 26,
+                            "wednesday": 27,
+                            "thursday": 28,
+                            "friday": 29,
+                            "saturday": 30
+                        },
+                        {
+                            "sunday": 31,
+                            "monday": null,
+                            "tuesday": null,
+                            "wednesday": null,
+                            "thursday": null,
+                            "friday": null,
+                            "saturday": null
+                        }
+                    ]
+                "#,
+        )]);
+        let answers_for_rows = ["3", "42", "91", "140", "189", "0"];
```

Comment:
For both this test and the test below, shouldn't the last value be 31? 🤔

Reply:
Yes, this is something that I wanted to bring up. It's a result of the way that the `std::ops::Add` trait is implemented for the `Primitive` type. The implementation is a little strange becuase the addition operation isn't actually defined for all possible instances of the `Primitive` type and the signature of add makes it difficult to propogate an error back to calllers. As a result, the add operation just returns 0 in the invalid cases. This leads to some weird behaviour when summing lists: 
```
echo [1 2 3] | sum           // Outputs 6
echo [1 2 3 null] | sum      // Outputs 0
echo [1 null 2 null 3]       // Outputs 3
```
My intuition about what should actually happen in these cases is that the add operation should return an error in the case that two incompatible `Primitive`s are added. I think that it would be fairly straightforward to remove the `std::ops::Add` trait implementation for the `Primitive` type and add an implementation for something like a `SafeAdd` trait where we can handle the case of invalid invocations of add. However, this would involve going through the code base and handling the `Result` anywhere else that `Primitive`s are added together. I'm not certain what the magnitude of that task would be. 

We could also implement the `std::ops::Add` trait such that invalid `Primitive`s (e.g. Strings) are interpreted as zero for the purposes of computing a sum, so the above examples would yield the following: 
```
echo [1 2 3] | sum           // Outputs 6
echo [1 2 3 null] | sum      // Outputs 6
echo [1 null 2 null 3]       // Outputs 6
```
Wanted to get some input from someone who has a little more knowledge wrt the semantics of the pipeline before I made any of the changes I mentioned above.

Reply:
I think propagating errors when we can't coerce would definitely be worth the effort (not for this PR, but another). Maybe we should implement `std::ops::Add` on the `UntaggedValue`, which has a more monadic like structure:

- can coerce: `a + b = c`
- can't coerce: `a + b = error`
- `error + b = error`

Now, for this PR, what can we do? I think the problem is that our impl for `Add` doesn't take `Nothing` into account. Since our `num_traits::Zero` implementation says `Nothing` is a zero value, we should adhere to the mathematical definition of that: `0 + a = a`.

Reply:
@DangerFunPants What do you think of ☝️ ?

Reply:
Okay, sounds good, I'll do that in the next little bit. For what it's worth this is what I was thinking wrt propogating errors: 

`UntaggedValue` already has an `std::ops::Add` implementation, we could define what it means to add `Primitive`s in the `UntaggedValue#add` implementation (i.e. move the code from `Primitive#add` to `UntaggedValue#add`). We could then return the error variant of `UntaggedValue` in the case that someone tries to compute a sum that isn't defined. This would mean that you could no longer add `Primitive`s without first wrapping them in an `UntaggedValue`. I think the best solution would be to have the type of the `add` function actually reflect what could take place (i.e. adding two `Primitive`s returns either a primitive or an error) by returning a `Result`.

Reply:
Yep, that sounds like a good approach. Feel free to open an issue (or a PR, if you have time to work on it).

The only other addition I'd make is that it feels weird if `a + b` would return an `Option` / `Result`, because then we could never write `a + b + c`. I'd probably drop trying to `impl std::ops::Add` and maybe just have a `try_add` method 🙂

Reply:
I thought a little more about use cases and I think it makes more sense to sum tables on a column-by-column basis, given that columns tend to represent a vector of values from the same domain (i.e. a vector of file sizes or a vector of durations), whereas the values contained in a row tend to be heterogeneous. Summing on a column-by-column basis has a number of obvious use cases.
1. Compute the sizes of all the files in the current directory:
```nu
du | select apparent physical | sum
```
2. Compute the sum of various metrics across all running processes:
```nu
ps | select cpu mem virtual | sum
```
Whereas computing the sum of a row rarely makes sense. As such, I've made a few more commits to implement `sum` on a column-by-column basis and update the tests to reflect this change.

Reply:
Sorry for the delay, @DangerFunPants . I'll try to 👀 on this soon, but I think it's reasonable to have columns summed. Not sure the implications of that decision, but it'd be nice to land it and see what kind of feedback we get. We're still pre-1.0, so we have the flexibility to easily go back and forth with some decisions once we learn more about the impact they have 🙂